File: examples/python/s3/main.py
import json
import urllib.parse


def handler(event, context):
    """
    Handles S3 object event invocations.

    event shape (events.S3Event):
    {
        "Records": [
            {
                "eventVersion": "2.1",
                "eventSource": "aws:s3",
                "awsRegion": "us-east-1",
                "eventTime": "2024-01-01T00:00:00Z",
                "eventName": "s3:ObjectCreated:Put",
                "s3": {
                    "bucket": {"name": "my-uploads-bucket", "arn": "arn:aws:s3:::my-uploads-bucket"},
                    "object": {"key": "images%2Fphoto.jpg", "eTag": "...", "sequencer": "..."}
                }
            }
        ]
    }
    """
    records = event.get("Records", [])
    print(f"Received {len(records)} S3 event(s)")

    for record in records:
        event_name = record.get("eventName", "")
        bucket = record["s3"]["bucket"]["name"]
        key = urllib.parse.unquote_plus(record["s3"]["object"]["key"])
        etag = record["s3"]["object"].get("eTag", "")

        print(f"Event: {event_name} | Bucket: {bucket} | Key: {key} | ETag: {etag}")

        if event_name.startswith("s3:ObjectCreated"):
            print(f"New object uploaded: s3://{bucket}/{key} — processing...")
            # Put your image/file processing logic here.

        elif event_name.startswith("s3:ObjectRemoved"):
            print(f"Object deleted: s3://{bucket}/{key} — cleaning up...")
            # Put your cleanup logic here.

    return {"status": "ok", "processed": len(records)}

File: examples/go/s3/main.go
package main

import (
	"context"
	"fmt"
	"log"
	"net/url"
	"strings"

	"github.com/aws/aws-lambda-go/events"
	"github.com/aws/aws-lambda-go/lambda"
)

// Handler processes S3 object events (creates, updates, deletes).
func Handler(ctx context.Context, event events.S3Event) (map[string]interface{}, error) {
	log.Printf("Received %d S3 event(s)", len(event.Records))

	processed := 0
	for _, record := range event.Records {
		bucket := record.S3.Bucket.Name
		key, err := url.QueryUnescape(record.S3.Object.Key)
		if err != nil {
			key = record.S3.Object.Key
		}
		etag := record.S3.Object.ETag
		eventName := record.EventName

		log.Printf("Event: %s | s3://%s/%s (ETag: %s)", eventName, bucket, key, etag)

		switch {
		case strings.HasPrefix(eventName, "s3:ObjectCreated"):
			log.Printf("New object uploaded: s3://%s/%s — processing...", bucket, key)
			// Put your object processing logic here — thumbnail generation,
			// virus scanning, metadata extraction, etc.

		case strings.HasPrefix(eventName, "s3:ObjectRemoved"):
			log.Printf("Object deleted: s3://%s/%s — cleaning up...", bucket, key)
			// Put your cleanup logic here — remove derived objects, update index, etc.
		}
		processed++
	}

	return map[string]interface{}{
		"status":    "ok",
		"processed": fmt.Sprintf("%d", processed),
	}, nil
}

func main() {
	lambda.Start(Handler)
}
